internal/blockchain/ethereum/ethereum.go

```diff
@@ -223,7 +236,37 @@ func (e *Ethereum) SetOperationHandler(namespace string, handler core.OperationC
 	e.callbacks.SetOperationalHandler(namespace, handler)
 }
 
+func (e *Ethereum) startBackground() {
+	batchSize := e.ethconnectConf.GetUint(EthconnectConfigBatchSize)
+	batchTimeout := uint(e.ethconnectConf.GetDuration(EthconnectConfigBatchTimeout).Milliseconds())
+
+	_ = e.backgroundRetry.Do(e.ctx, fmt.Sprintf("ethereum connector %s", e.Name()), func(attempt int) (retry bool, err error) {
+		stream, err := e.streams.ensureEventStream(e.ctx, e.topic, batchSize, batchTimeout)
+		if err != nil {
+			return true, err
+		}
+
+		e.streamID = stream.ID
+		log.L(e.ctx).Infof("Event stream: %s (topic=%s)", e.streamID, e.topic)
+
+		e.closed = make(chan struct{})
+		go e.eventLoop()
```

Comment:
Looks like we're starting the Go routine, before we've attempted the `e.wsconn.Connect` - so each time we go round the loop. But we're not ensuring it's cancelled if we fail. So I'm thinking it's possible we'll leak go routines.

Reply:
Correct, great catch - we will run multiple go routines until it's successful and leak them! I was relying on the context being cancelled since each event loop will listen to that but not if we successfully go through, will make sure to amend that